### ISSUE TITLE
Allow extension (merging) of keybidings

### ISSUE THREAD

**Issue by spencerlyon2** [dvcrn/proton#21]
We should allow layers to extend keybindings defined in core or other layers.

Here's an example layer that could utilize this:

```
(ns proton.layers.pomodoro.core
  (:use [proton.layers.base :only [init-layer! get-initial-config get-keybindings get-packages get-keymaps]]))

(defmethod get-initial-config :pomodoro
  []
  [])

(defmethod init-layer! :pomodoro
  [_ config]
  (println "init pomodoro"))

(defmethod get-keybindings :pomodoro
  []
  {:t {:category "toggles"}
     :p {:category "pomodoro"
         :s {:action "pomodoro:start"
             :title "Start Timer"}
         :a {:action "pomodoro:abort"
             :title "Abort Timer"}}})

(defmethod get-packages :pomodoro
  []
  [:pomodoro])

(defmethod get-keymaps :pomodoro
  []
  [])
```

What I am going for here is to add another category under the `t` chain defined in core so that I can toggle my pomodoro timer. So what I'd like to see is that `SPC t` has all the existing options, but now has one more option for `pomodoro` that will drop me one layer lower in the chain and have options for `a` and `s`. 

Right now, this is now what I'm seeing. Here are some screenshots to demonstrate the problem:

When I do just `SPC` I get this:

![screen shot 2015-12-08 at 11 08 26 am](https://cloud.githubusercontent.com/assets/1001948/11660663/274fe398-9d9c-11e5-9e32-c68c10d73139.png)

When I do `SPC t` it appears that everything was wiped out:

![screen shot 2015-12-08 at 11 08 53 am](https://cloud.githubusercontent.com/assets/1001948/11660670/3343dd08-9d9c-11e5-814e-39d29b50717e.png)

It also appears that `SPC p` has replaced the "project" section from core with my pomodoro section defined here:

![screen shot 2015-12-08 at 11 08 59 am](https://cloud.githubusercontent.com/assets/1001948/11660681/46e880d4-9d9c-11e5-985a-c1d58686250a.png)

**Comment by dvcrn** [comment]
Hi,
```
  {:t {:category "toggles"}
     :p {:category "pomodoro"
         :s {:action "pomodoro:start"
             :title "Start Timer"}
         :a {:action "pomodoro:abort"
             :title "Abort Timer"}}})
```

here you are just overwriting the `:t` key: `:t {:category "toggles"}` (the `:t` map is closed). I think you have a little bracket misplacement.

Nevertheless, code in question is [here](https://github.com/dvcrn/proton/blob/master/src/proton/lib/proton.cljs#L29-L30). Looks like I misunderstood what `merge` does. Need to apply a recursive merge here that goes into sub-maps

**Comment by dvcrn** [comment]
Found some deep-merge implementations and added them as helpers [here](https://github.com/dvcrn/proton/commit/99e31aaae639fa4e894d9d824b5d3af827a02164). Didn't have time yet to check deep-merging of keybindings

**Comment by spencerlyon2** [comment]
So the deep merge seems to work.

How could I extend the `:t` map (i.e. what did you mean by "it is closed")?

**Comment by dvcrn** [comment]
I mean, you want `:p` inside `:t`, don't you? So `:t` should be a element of the map thats assigned to `:p`. 

In your example you define 2 root keys: `:t` and `:p`. 
```
{:t {:category "toggles"}
   :p {:category "pomodoro"
       :s {:action "pomodoro:start"
           :title "Start Timer"}
       :a {:action "pomodoro:abort"
           :title "Abort Timer"}}}
```

should be
```
{:t {:category "toggles"
     :p {:category "pomodoro"
         :s {:action "pomodoro:start"
             :title "Start Timer"}
         :a {:action "pomodoro:abort"
             :title "Abort Timer"}}}}
```

now `:t` is inside `:p` and the chain would be `p -> t -> s` for pomodoro start

**Comment by spencerlyon2** [comment]
:facepalm: oh man that was a dumb mistake. Parifiner is cool, but I'm going to have to get used to it adding brackets in places.

Anyway, this is closed by #28

**Comment by dvcrn** [comment]
Haha that's what I thought :) 

You technically never really have to touch your brackets with Parinfer. Just indent it so the `:p` aligns with `:category` and it will change the brackets for you.
